extension/chrome/elements/attachment.ts

```diff
@@ -320,10 +320,71 @@ export class AttachmentDownloadView extends View {
     });
     Xss.sanitizeRender(this.downloadButton, this.originalButtonHTML || '');
     if (result.success) {
+      const attachmentForSave = new Attachment({ name: result.filename, type: this.attachment.type, data: result.content });
       if (!result.filename || ['msg.txt', 'null'].includes(result.filename)) {
         result.filename = this.attachment.name;
       }
-      Browser.saveToDownloads(new Attachment({ name: result.filename, type: this.attachment.type, data: result.content }));
+      const blacklistedFiles = [
+        '.ade',
+        '.adp',
+        '.apk',
+        '.appx',
+        '.appxbundle',
+        '.bat',
+        '.cab',
+        '.chm',
+        '.cmd',
+        '.com',
+        '.cpl',
+        '.diagcab',
+        '.diagcfg',
+        '.diagpack',
+        '.dll',
+        '.dmg',
+        '.ex',
+        '.ex_',
+        '.exe',
+        '.hta',
+        '.img',
+        '.ins',
+        '.iso',
+        '.isp',
+        '.jar',
+        '.jnlp',
+        '.js',
+        '.jse',
+        '.lib',
+        '.lnk',
+        '.mde',
+        '.msc',
+        '.msi',
+        '.msix',
+        '.msixbundle',
+        '.msp',
+        '.mst',
+        '.nsh',
+        '.pif',
+        '.ps1',
+        '.scr',
+        '.sct',
+        '.shb',
+        '.sys',
+        '.vb',
+        '.vbe',
+        '.vbs',
+        '.vhd',
+        '.vxd',
+        '.wsc',
+        '.wsf',
+        '.wsh',
+        '.xll',
+      ];
+      if (blacklistedFiles.some(badFileExtension => result.filename?.endsWith(badFileExtension))) {
+        const badFileExtensionWarning = 'This executable file was not checked for viruses, and may be dangerous to download or run. Proceed anyway?'; // xss-safe-value
+        BrowserMsg.send.showWarningForAttachmentDownload(this.parentTabId, { message: badFileExtensionWarning, attachment: attachmentForSave });
+        return;
+      }
+      Browser.saveToDownloads(attachmentForSave);
```

Comment:
repeated codes - could be in a function somewhere in `extension/js/common/core/attachments`?